File: tests/codegen/foreign-object-foldm.mjs
// Original JS _foldM vs real Object/ST/Exception FFI and Rc/Arc runtime.
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { mkdirSync, mkdtempSync, readFileSync, writeFileSync } from 'node:fs';
import { join, resolve, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';
import { codegenPrelude } from '../../output/Purust.CodeGen/index.js';
import { fromFoldable } from '../../output/Data.Set/index.js';
import { ordString } from '../../output/Data.Ord/index.js';
import { foldableArray } from '../../output/Data.Foldable/index.js';
import { threadedPrelude, threadedRust } from '../../src/Purust/Threading.js';
import { rustStringLiteral } from '../../src/Purust/Utf16.js';

const root = fileURLToPath(new URL('../../', import.meta.url));
const mount = resolve(root, '../../b8x/run/bak');
const directory = mkdtempSync(join(mount, 'rust/output/purust-object-foldm-'));
const remote = path => '/var/www/b8x/run/bak/' + relative(mount, path);
const read = path => readFileSync(resolve(root, path), 'utf8');
const hash = path => createHash('sha256').update(readFileSync(path)).digest('hex');
const report = { complete: false, commands: [], modes: [], inputs: [
  'output/Purust.CodeGen/index.js', 'src/Purust/RecordFields.js', 'src/Purust/Utf16.js', 'src/Purust/Threading.js',
  'tests/codegen/foreign-object-foldm.mjs', 'tests/codegen/fixtures/foreign-object-foldm.rs',
  'tests/runtime/perceus_ptr/src/lib.rs', 'tests/runtime/perceus_ptr/src/local.rs', 'tests/runtime/perceus_ptr/src/threaded.rs',
  '../purust-exceptions/src/Effect/Exception.rs', '../purust-foreign-object/src/Foreign/Object.rs',
  '../purust-foreign-object/src/Foreign/Object/ST.rs', '../purust-foreign-object/src/Foreign/Object.js',
].map(p => { const path = resolve(root, p); return { path, sha256: hash(path) }; }) };
const save = () => writeFileSync(join(directory, 'report.json'), JSON.stringify(report, null, 2));
console.log(directory);
const source = read('../purust-foreign-object/src/Foreign/Object.js');
const { _foldM } = await import('data:text/javascript;base64,' + Buffer.from(source).toString('base64'));
const immediate = z => next => next(z), deferred = action => next => () => next(action())();
const hex = value => Array.from({ length: value.length }, (_, i) => value.charCodeAt(i).toString(16).padStart(4, '0')).join('');
const describe = value => typeof value === 'string' ? 's:' + hex(value) : String(value);
const collect = z => k => v => z + hex(k) + '=' + describe(v) + ';';
const own = entries => {
  const object = Object.create({ inherited: 999 });
  for (const [key, value] of entries) Object.defineProperty(object, key, { value, writable: true, configurable: true, enumerable: true });
  return object;
};
const cases = [[]];
const keys = ['z', '10', '2', '01', '4294967294', '4294967295', '-1', '1.0', 'NaN', '__proto__', 'constructor',
  'hasOwnProperty', 'toString', '', '\0', 'é', '😀', '\ud800', '\udfff', 'a\ud800b'];
const values = [undefined, null, true, false, 0, -42, '', 'é😀', '\ud800a\udfff'];
cases.push(keys.map((key, i) => [key, values[i % values.length]]));
for (const value of values) cases.push([['value', value]]);
let state = 0x8fa1543c;
const random = () => { state ^= state << 13; state ^= state >>> 17; state ^= state << 5; return state >>> 0; };
for (let n = 0; n < 256; n++) cases.push(Array.from({ length: n % 24 }, () => [keys[random() % keys.length], values[random() % values.length]]));
const reference = { vectors: cases.map(entries => _foldM(immediate)(collect)('seed;')(own(entries))) };
const initial = {}; assert.strictEqual(_foldM(() => { throw new Error('empty bind'); })(() => { throw new Error('empty callback'); })(initial)({}), initial);
const eagerObject = { a: 1, b: 2, c: 3 };
reference.eager = _foldM(immediate)(z => k => v => {
  if (k === 'a') { eagerObject.b = 20; delete eagerObject.c; eagerObject.d = 4; }
  return collect(z)(k)(v);
})('seed;')(eagerObject);
const lazyObject = { a: 1, b: 2, c: 3 }; let lazyCalls = 0;
const lazy = _foldM(deferred)(z => k => v => { lazyCalls++; return () => collect(z)(k)(v); })(() => 'seed;')(lazyObject);
assert.equal(lazyCalls, 0); lazyObject.b = 20; delete lazyObject.c; lazyObject.d = 4;
reference.deferredFirst = lazy(); lazyObject.a = 10; lazyObject.c = 30; reference.deferredSecond = lazy(); assert.equal(lazyCalls, 6);
const repeatedObject = { a: 1 };
reference.repeated = _foldM(z => next => { const first = next(z); repeatedObject.a = 2; return [first, next(z)]; })(collect)('seed;')(repeatedObject);
let shortBinds = 0; const marker = [];
assert.strictEqual(_foldM(z => () => { shortBinds++; return z; })(() => { throw new Error('short callback'); })(marker)({ a: null, b: undefined }), marker);
reference.shortBinds = shortBinds;
const child = {}; assert.strictEqual(_foldM(immediate)(() => () => value => value)(null)({ payload: child }), child);
for (const isLazy of [false, true]) {
  const error = new TypeError('fold callback'); let calls = 0;
  assert.throws(() => {
    const result = _foldM(isLazy ? deferred : immediate)(() => () => () => { calls++; throw error; })(isLazy ? () => undefined : undefined)({ a: 1, b: 2 });
    if (isLazy) result();
  }, caught => caught === error); assert.equal(calls, 1);
}
const bindError = new Error('bind failure');
assert.throws(() => _foldM(() => { throw bindError; })(() => { throw new Error('callback'); })(null)({ a: 1 }), error => error === bindError);
writeFileSync(join(directory, 'reference.json'), JSON.stringify(reference, null, 2));
report.jsReference = { passed: true, vectors: cases.length, eager: reference.eager,
  deferred: [reference.deferredFirst, reference.deferredSecond], repeated: reference.repeated, shortBinds };
save();
const rustValue = value => value === undefined ? 'Value::Unit' : value === null ? 'Value::Null' :
  typeof value === 'boolean' ? `Value::Bool(${value})` : typeof value === 'number' ? `mk_int(${value})` : `Value::String(${rustStringLiteral(value)})`;
const vectors = `#[test]\nfn javascript_own_property_vectors() {\n${cases.map((entries, i) => `
    let object=Rc::new(Object::from_entries(vec![${entries.map(([k,v]) => `(${rustStringLiteral(k)},${rustValue(v)})`).join(',')}]));
    assert_eq!(fold_m(immediate(),Func3::Static(collect),mk_string("seed;"),object).unwrap_string(),${JSON.stringify(reference.vectors[i])},"vector ${i}");`).join('\n')}\n}`;
const constants = `const EXPECTED_EAGER:&str=${JSON.stringify(reference.eager)};
const EXPECTED_DEFERRED_FIRST:&str=${JSON.stringify(reference.deferredFirst)};
const EXPECTED_DEFERRED_SECOND:&str=${JSON.stringify(reference.deferredSecond)};
const EXPECTED_REPEATED:&[&str]=&${JSON.stringify(reference.repeated)};
const EXPECTED_SHORT_CIRCUIT_BINDS:usize=${reference.shortBinds};\n`;
const prelude = codegenPrelude(fromFoldable(foldableArray)(ordString)(['']));
try {
  for (const threaded of [false, true]) {
    const mode = threaded ? 'threaded' : 'normal', dir = join(directory, mode); mkdirSync(dir);
    const adapt = source => threaded ? threadedRust(source) : source;
    const code = `${threaded ? threadedPrelude(prelude) : prelude}
extern crate self as purust_core;
extern crate self as Purs_Foreign_Object_ST;
mod perceus_ptr { ${read('tests/runtime/perceus_ptr/src/lib.rs').replace('mod local;', `mod local { ${read('tests/runtime/perceus_ptr/src/local.rs')} }`).replace('mod threaded;', `mod threaded { ${read('tests/runtime/perceus_ptr/src/threaded.rs')} }`)} }
mod object_st { ${adapt(read('../purust-foreign-object/src/Foreign/Object/ST.rs'))} }
pub use object_st::STObject;
mod object { ${adapt(read('../purust-foreign-object/src/Foreign/Object.rs'))} }
use object::Object;
mod exception { ${adapt(read('../purust-exceptions/src/Effect/Exception.rs'))} }
${constants}${adapt(read('tests/codegen/fixtures/foreign-object-foldm.rs'))}
${adapt(vectors)}`;
    writeFileSync(join(dir, 'checks.rs'), code);
    writeFileSync(join(dir, 'Cargo.toml'), '[package]\nname="object_foldm_ffi"\nversion="0.0.0"\nedition="2021"\n[lib]\npath="checks.rs"\n[features]\nthreaded=[]\n');
    const args = ['exec', '-w', remote(dir), '-e', 'CARGO_BUILD_JOBS=1', '-e', 'CARGO_PROFILE_DEV_DEBUG=0', '-e', 'CARGO_PROFILE_TEST_DEBUG=0',
      '-e', 'CARGO_INCREMENTAL=0', 'core-api-cli-1', 'cargo', 'test', '--offline', '--quiet', ...(threaded ? ['--features', 'threaded'] : []), '--lib', '--', '--test-threads=1'];
    const result = spawnSync('docker', args, { encoding: 'utf8', timeout: 90000, maxBuffer: 8 * 1024 * 1024 });
    writeFileSync(join(directory, `test-${mode}.json`), JSON.stringify({ args, status: result.status, error: result.error?.message,
      stdout: result.stdout, stderr: result.stderr }, null, 2));
    report.commands.push({ mode, status: result.status }); save();
    assert.equal(result.status, 0, `${mode}: ${result.error ?? ''}\n${result.stdout}\n${result.stderr}`);
    // The local perceus_ptr source also contributes four native unit tests.
    const foldTests = threaded ? 11 : 10, runtimeTests = threaded ? 0 : 4;
    const expected = foldTests + runtimeTests;
    assert.match(result.stdout, new RegExp(`${expected} passed; 0 failed`));
    report.modes.push({ mode, tests: expected, foldTests, runtimeTests, vectors: cases.length, concurrentReplays: threaded ? 512 : 0 }); save();
    console.log(`${mode}: ${expected} tests, ${cases.length} original JS vectors passed`);
  }
  report.inputs.forEach(input => assert.equal(hash(input.path), input.sha256, `input changed: ${input.path}`));
  report.complete = true;
} finally { save(); }

File: tests/codegen/fixtures/foreign-object-foldm.rs
use object::Foreign_Object__foldM as fold_m;
use std::rc::Rc;
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    Mutex,
};

fn native(entries: Vec<(&str, Value)>) -> Rc<Object> {
    Rc::new(Object::from_entries(
        entries.into_iter().map(|(k, v)| (k.into(), v)).collect(),
    ))
}
fn immediate() -> Func2<Value, Func1<Value, Value>, Value> {
    Func2::Static(|value, next| next(value))
}
fn effect(value: Value) -> Value {
    Value::Func1(Func1::Shared(Rc::new(move |_| value.clone())))
}
fn deferred() -> Func2<Value, Func1<Value, Value>, Value> {
    Func2::Static(|action, next| {
        Value::Func1(Func1::Shared(Rc::new(move |_| {
            next(action.unwrap_func1()(Value::Unit)).unwrap_func1()(Value::Unit)
        })))
    })
}
fn hex(value: &str) -> String {
    purust_string_to_utf16(value)
        .iter()
        .map(|u| format!("{u:04x}"))
        .collect()
}
fn describe(value: &Value) -> String {
    match value.resolve() {
        Value::Unit => "undefined".into(),
        Value::Null => "null".into(),
        Value::Bool(v) => v.to_string(),
        Value::Int(v) => v.to_string(),
        Value::String(v) => format!("s:{}", hex(v)),
        _ => panic!("unsupported differential payload"),
    }
}
fn collect(accumulator: Value, key: String, value: Value) -> Value {
    Value::String(format!(
        "{}{}={};",
        accumulator.unwrap_string(),
        hex(&key),
        describe(&value)
    ))
}

#[test]
fn empty_fold_retains_initial_identity_without_bind_or_callback() {
    let handle = Rc::new(Object::empty());
    let initial = Value::Class(Rc::new(handle.clone()));
    let result = fold_m(
        Func2::Static(|_, _| panic!("empty bind")),
        Func3::Static(|_, _, _| panic!("empty callback")),
        initial,
        native(vec![]),
    );
    assert!(Rc::ptr_eq(&handle, result.unwrap_class::<Rc<Object>>()));
}

#[test]
fn eager_bind_observes_replacements_skips_deletions_and_ignores_additions() {
    let object = native(vec![("a", mk_int(1)), ("b", mk_int(2)), ("c", mk_int(3))]);
    let changing = object.clone();
    let result = fold_m(
        immediate(),
        Func3::Shared(Rc::new(move |z, k, v| {
            if k == "a" {
                changing.insert("b".into(), mk_int(20));
                changing.remove("c");
                changing.insert("d".into(), mk_int(4));
            }
            collect(z, k, v)
        })),
        mk_string("seed;"),
        object,
    );
    assert_eq!(result.unwrap_string(), EXPECTED_EAGER);
}

#[test]
fn deferred_bind_reads_on_continuation_execution_and_replays() {
    let object = native(vec![("a", mk_int(1)), ("b", mk_int(2)), ("c", mk_int(3))]);
    let calls = Rc::new(AtomicUsize::new(0));
    let seen = calls.clone();
    let action = fold_m(
        deferred(),
        Func3::Shared(Rc::new(move |z, k, v| {
            seen.fetch_add(1, Ordering::SeqCst);
            effect(collect(z, k, v))
        })),
        effect(mk_string("seed;")),
        object.clone(),
    );
    assert_eq!(calls.load(Ordering::SeqCst), 0);
    object.insert("b".into(), mk_int(20));
    object.remove("c");
    object.insert("d".into(), mk_int(4));
    assert_eq!(
        action.unwrap_func1()(Value::Unit).unwrap_string(),
        EXPECTED_DEFERRED_FIRST
    );
    object.insert("a".into(), mk_int(10));
    object.insert("c".into(), mk_int(30));
    assert_eq!(
        action.unwrap_func1()(Value::Unit).unwrap_string(),
        EXPECTED_DEFERRED_SECOND
    );
    assert_eq!(calls.load(Ordering::SeqCst), 6);
}

#[test]
fn one_continuation_can_run_multiple_times_with_current_values() {
    let object = native(vec![("a", mk_int(1))]);
    let changing = object.clone();
    let result = fold_m(
        Func2::Shared(Rc::new(move |z, next| {
            let first = next(z.clone());
            changing.insert("a".into(), mk_int(2));
            mk_array(vec![first, next(z)])
        })),
        Func3::Static(collect),
        mk_string("seed;"),
        object,
    )
    .unwrap_array();
    assert_eq!(
        result.iter().map(Value::unwrap_string).collect::<Vec<_>>(),
        EXPECTED_REPEATED
    );
}

#[test]
fn bind_may_short_circuit_without_forcing_callback() {
    let calls = Rc::new(AtomicUsize::new(0));
    let seen = calls.clone();
    let original = Rc::new(vec![mk_int(42)]);
    let result = fold_m(
        Func2::Shared(Rc::new(move |z, _| {
            seen.fetch_add(1, Ordering::SeqCst);
            z
        })),
        Func3::Static(|_, _, _| panic!("short-circuited callback")),
        Value::Array(original.clone()),
        native(vec![("a", Value::Null), ("b", Value::Unit)]),
    );
    assert!(Rc::ptr_eq(&original, &result.unwrap_array()));
    assert_eq!(calls.load(Ordering::SeqCst), EXPECTED_SHORT_CIRCUIT_BINDS);
}

#[test]
fn foreign_object_payload_identity_is_not_cloned_into_a_new_object() {
    let payload = Rc::new(Object::empty());
    let expected = payload.clone();
    let result = fold_m(
        immediate(),
        Func3::Shared(Rc::new(move |_, key, value| {
            assert_eq!(key, "payload");
            assert!(Rc::ptr_eq(&expected, value.unwrap_class::<Rc<Object>>()));
            value
        })),
        Value::Unit,
        native(vec![("payload", Value::Class(Rc::new(payload.clone())))]),
    );
    assert!(Rc::ptr_eq(&payload, result.unwrap_class::<Rc<Object>>()));
}

#[test]
fn continuation_owns_record_until_dropped_without_leaking() {
    let object = native(vec![("a", mk_int(1))]);
    let weak = Rc::downgrade(&object);
    let continuation = fold_m(
        Func2::Static(|_, next| Value::Func1(next)),
        Func3::Static(collect),
        mk_string("seed;"),
        object.clone(),
    );
    drop(object);
    assert!(weak.upgrade().is_some());
    assert_eq!(
        continuation.unwrap_func1()(mk_string("seed;")).unwrap_string(),
        "seed;0061=1;"
    );
    drop(continuation);
    assert!(weak.upgrade().is_none());
}

#[test]
fn original_effect_exception_is_propagated_and_stops_execution() {
    use exception::*;
    for lazy in [false, true] {
        let error = Effect_Exception_errorWithName("fold callback".into(), "TypeError".into());
        let raised = error.clone();
        let calls = Rc::new(AtomicUsize::new(0));
        let seen = calls.clone();
        let result = purust_exception_try(|| {
            let callback = Func3::Shared(Rc::new(move |_, _, _| {
                seen.fetch_add(1, Ordering::SeqCst);
                purust_exception_raise(raised.clone())
            }));
            let result = fold_m(
                if lazy { deferred() } else { immediate() },
                callback,
                if lazy {
                    effect(Value::Unit)
                } else {
                    Value::Unit
                },
                native(vec![("a", mk_int(1)), ("b", mk_int(2))]),
            );
            if lazy {
                result.unwrap_func1()(Value::Unit)
            } else {
                result
            }
        });
        let caught = match result {
            Err(error) => error,
            Ok(_) => panic!("swallowed callback error"),
        };
        assert!(std::sync::Arc::ptr_eq(
            &purust_exception_unbox(&error),
            &purust_exception_unbox(&caught)
        ));
        assert_eq!(calls.load(Ordering::SeqCst), 1);
    }
}

#[test]
fn original_bind_exception_is_not_replaced() {
    use exception::*;
    let error = Effect_Exception_error("bind failure".into());
    let raised = error.clone();
    let result = purust_exception_try(|| {
        fold_m(
            Func2::Shared(Rc::new(move |_, _| purust_exception_raise(raised.clone()))),
            Func3::Static(|_, _, _| panic!("callback after failed bind")),
            Value::Unit,
            native(vec![("a", mk_int(1))]),
        )
    });
    let caught = match result {
        Err(error) => error,
        Ok(_) => panic!("swallowed bind error"),
    };
    assert!(std::sync::Arc::ptr_eq(
        &purust_exception_unbox(&error),
        &purust_exception_unbox(&caught)
    ));
}

#[cfg(feature = "threaded")]
#[test]
fn arc_continuation_can_be_replayed_across_threads() {
    let action = fold_m(
        deferred(),
        Func3::Static(|z, k, v| effect(collect(z, k, v))),
        effect(mk_string("seed;")),
        native(vec![("a", mk_int(1)), ("b", mk_int(2))]),
    );
    let workers = (0..8)
        .map(|_| {
            let action = action.clone();
            std::thread::spawn(move || {
                for _ in 0..64 {
                    assert_eq!(
                        action.unwrap_func1()(Value::Unit).unwrap_string(),
                        "seed;0061=1;0062=2;"
                    );
                }
            })
        })
        .collect::<Vec<_>>();
    for worker in workers {
        worker.join().unwrap();
    }
}
